tools/reduce/reduce.cpp

```diff
@@ -0,0 +1,229 @@
+                       spvSoftwareVersionDetailsString());
+        return {REDUCE_STOP, 0};
+      } else if (0 == strcmp(cur_arg, "--help") || 0 == strcmp(cur_arg, "-h")) {
+        PrintUsage(argv[0]);
+        return {REDUCE_STOP, 0};
+      } else if ('\0' == cur_arg[1]) {
+        // We do not support reduction from standard input.  We could support
+        // this if there was a compelling use case.
+        PrintUsage(argv[0]);
+        return {REDUCE_STOP, 0};
+      } else if (0 == strncmp(cur_arg,
+                              "--step-limit=", sizeof("--step-limit=") - 1)) {
+        const auto split_flag = spvtools::utils::SplitFlagArgs(cur_arg);
+        char* end = nullptr;
+        errno = 0;
+        const auto step_limit =
+            static_cast<uint32_t>(strtol(split_flag.second.c_str(), &end, 10));
+        assert(end != split_flag.second.c_str() && errno == 0);
+        reducer_options->set_step_limit(step_limit);
+      }
+    } else if (positional_arg_index == 0) {
+      // Input file name
+      assert(!*in_file);
+      *in_file = cur_arg;
+      positional_arg_index++;
+    } else if (positional_arg_index == 1) {
+      assert(!*interestingness_test);
+      *interestingness_test = cur_arg;
+      positional_arg_index++;
+    } else {
+      spvtools::Error(ReduceDiagnostic, nullptr, {},
+                      "Too many positional arguments specified");
+      return {REDUCE_STOP, 1};
+    }
+  }
+
+  if (!*in_file) {
+    spvtools::Error(ReduceDiagnostic, nullptr, {}, "No input file specified");
+    return {REDUCE_STOP, 1};
+  }
+
+  if (!*interestingness_test) {
+    spvtools::Error(ReduceDiagnostic, nullptr, {},
+                    "No interestingness test specified");
+    return {REDUCE_STOP, 1};
+  }
+
+  return {REDUCE_CONTINUE, 0};
+}
+
+}  // namespace
+
+const auto kDefaultEnvironment = SPV_ENV_UNIVERSAL_1_3;
+
+int main(int argc, const char** argv) {
+  const char* in_file = nullptr;
+  const char* interestingness_test = nullptr;
+
+  spv_target_env target_env = kDefaultEnvironment;
+  spvtools::ReducerOptions reducer_options;
+
+  ReduceStatus status =
+      ParseFlags(argc, argv, &in_file, &interestingness_test, &reducer_options);
+
+  if (status.action == REDUCE_STOP) {
+    return status.code;
+  }
+
+  if (!CheckExecuteCommand()) {
+    std::cerr << "could not find shell interpreter for executing a command"
+              << std::endl;
+    return 2;
+  }
+
+  Reducer reducer(target_env);
+
+  reducer.SetInterestingnessFunction([&](std::vector<uint32_t> binary,
+          uint32_t reductions_applied) -> bool {
+    std::stringstream ss;
+    ss << "temp_" << std::setw(4) << std::setfill('0') << reductions_applied
+        << ".spv";
+    const auto spv_file = ss.str();
+    const std::string command = std::string(interestingness_test) + " "
+            + spv_file;
+    assert(WriteFile(spv_file.c_str(), "wb", &binary[0], binary.size()));
```

Comment:
Asserts are removed in the release version, so the file will not be written in the release version.  Better to split this into two statements.  One that writes and saves the return values.  Then assert on the returned value.

Reply:
Good catch!  Done.